File: main/binary-tree-inorder-traversal/binary-tree-inorder-traversal-fast2.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public IList<int> InorderTraversal(TreeNode root) {
        var ret = new List<int>();
        var stack = new Stack<TreeNode>();

        for (; ; ) {
            for (; root != null; root = root.left) stack.Push(root);

            if (stack.Count == 0) break;

            root = stack.Pop();
            ret.Add(root.val);
            root = root.right;
        }

        return ret;
    }
}

File: main/binary-tree-inorder-traversal/binary-tree-inorder-traversal-fast2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def inorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        stack = []
        while root is not None:
            stack.append(root)
            root = root.left

        ret = []
        while stack:
            root = stack.pop()
            ret.append(root.val)
            root = root.right

            while root is not None:
                stack.append(root)
                root = root.left

        return ret
